crates/ludtwig/src/check/rules/duplicate_block_names.rs

```diff
@@ -0,0 +1,86 @@
+use std::collections::HashMap;
+
+use ludtwig_parser::syntax::typed::{AstNode, TwigBlock};
+use ludtwig_parser::syntax::untyped::{SyntaxKind, SyntaxNode, WalkEvent};
+
+use crate::check::rule::{CheckResult, Rule, RuleExt, RuleRunContext, Severity};
+
+pub struct RuleDuplicateBlockNames;
+
+impl Rule for RuleDuplicateBlockNames {
+    fn name(&self) -> &'static str {
+        "duplicate-block-names"
+    }
+
+    fn check_root(&self, node: SyntaxNode, _ctx: &RuleRunContext) -> Option<Vec<CheckResult>> {
+        // keep track of some state during tree traversal
+        let mut block_table: HashMap<String, TwigBlock> = HashMap::new();
+
+        let mut check_results = vec![];
+        let mut tree_iter = node.preorder();
+        while let Some(walk) = tree_iter.next() {
+            match walk {
+                WalkEvent::Enter(element) => {
+                    if element.kind() == SyntaxKind::ERROR {
+                        tree_iter.skip_subtree(); // Skip everything under error nodes!
+                    }
+                    let Some(block) = TwigBlock::cast(element) else {
+                        continue;
+                    };
+                    let Some(name) = block.name() else {
+                        continue;
+                    };
+                    if let Some(first_definition) = block_table
+                        .get(name.text())
+                        .and_then(TwigBlock::starting_block)
+                    {
+                        check_results.push(
+                            self.create_result(Severity::Error, "block name duplicate")
+                                .primary_note(
+                                    name.text_range(),
+                                    format!("duplicate block '{}'", name.text()),
+                                )
+                                .secodary_note(
+                                    first_definition.syntax().text_range(),
+                                    "first defined here",
+                                ),
+                        );
+                    }
+                    block_table.insert(name.text().to_owned(), block);
+                }
+                WalkEvent::Leave(_element) => {}
+            }
+        }
+
+        if check_results.is_empty() {
+            None
+        } else {
+            Some(check_results)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use expect_test::expect;
+
+    use crate::check::rules::test::test_rule;
+
+    #[test]
+    fn rule_reports() {
+        test_rule(
+            "duplicate-block-names",
+            r#"{% block foo %}{% endblock %} {%block foo%} {% endblock %}"#,
```

Comment:
Happy to get some Ideas for edge cases

Reply:
This rule implementation is much cleaner than I expected, great job 👍 .
I don't know any specific edge cases to test, except maybe one:

You are iterating the tree by yourself and therefore must uphold the following invariants:
- Don't look into Error Syntax Nodes ✅ 
- Check for `ludtwig-ignore` statements before your inspected syntax node ❌ 
The latter is kinda stupid because why would you want to ignore this rule? But for completeness it should be possible to ignore it 😂 . For inspiration this is also done in the indentation rule (see function `check_for_rule_ignore_enter`)